specs/core/1_shard-data-chains.md

```diff
@@ -95,517 +98,403 @@ We define the following Python custom types for type hinting and readability:
+    seed = hash(get_seed(state, epoch) + int_to_bytes(slot, length=8) + int_to_bytes(shard, length=8))
+    compute_proposer_index(state, active_indices, seed)
 ```
 
-### `flatten_shard_header`
+### Shard state mutators
+
+#### `add_reward`
 
 ```python
-def flatten_shard_header(header: ShardBlockHeader) -> Bytes[SHARD_HEADER_SIZE]:
-    """
-    Converts a shard block header into a flat object with the same hash tree root. Used
-    in the crosslink construction.
-    """
-    committee_size = len(header.core.attester_bitfield)
-    attester_bits = [header.core.attester_bitfield[i] if i < committee_size else 0 for i in range(256)]
-    attester_bytes = bytes([sum([attester_bits[i + j] << j for j in range(8)]) for i in range(0, 256, 8)])
-    return (
-        pad(int_to_bytes(header.core.slot, length=8), 32) +
-        header.core.beacon_chain_root +
-        header.core.parent_root +
-        header.core.data_root +
-        header.core.state_root +
-        pad(int_to_bytes(header.core.total_bytes, length=8), 32) +
-        attester_bytes +
-        b'\x00' * 32 +
-        pad(header.signatures.attestation_signature, 128) +
-        pad(header.signatures.proposer_signature, 128)
-    )
+def add_reward(state: BeaconState, shard_state: ShardState, index: ValidatorIndex, delta: Gwei) -> None:
+    epoch = compute_epoch_of_shard_slot(state.slot)
+    older_committee = get_period_committee(state, shard_state.shard, compute_shard_period_start_epoch(epoch, 2))
+    newer_committee = get_period_committee(state, shard_state.shard, compute_shard_period_start_epoch(epoch, 1))
+    if index in older_committee:
+        shard_state.older_committee_rewards[older_committee.index(index)] += delta
+    elif index in newer_committee:
+        shard_state.newer_committee_rewards[newer_committee.index(index)] += delta
 ```
 
-### `compute_crosslink_data_root`
+#### `add_fee`
 
 ```python
-def compute_crosslink_data_root(blocks: Sequence[ShardBlock]) -> Hash:
-    header = b''.join([flatten_shard_header(get_shard_header(block)) for block in blocks])
-    footer = b''.join([block.core.data for block in blocks])
-    MAX_SIZE = SHARD_BLOCK_SIZE_LIMIT * SHARD_SLOTS_PER_BEACON_SLOT * SLOTS_PER_EPOCH * MAX_EPOCHS_PER_CROSSLINK
-    return hash_tree_root(BytesN[MAX_SIZE](pad(header + footer, MAX_SIZE)))
+def add_fee(state: BeaconState, shard_state: ShardState, index: ValidatorIndex, delta: Gwei) -> None:
+    epoch = compute_epoch_of_shard_slot(state.slot)
+    older_committee = get_period_committee(state, shard_state.shard, compute_shard_period_start_epoch(epoch, 2))
+    newer_committee = get_period_committee(state, shard_state.shard, compute_shard_period_start_epoch(epoch, 1))
+    if index in older_committee:
+        shard_state.older_committee_fees[older_committee.index(index)] += delta
+    elif index in newer_committee:
+        shard_state.newer_committee_fees[newer_committee.index(index)] += delta
```

Comment:
`add_reward` and `add_fee` seem very mergeable. But this just brings us into the larger question of whether or not we can merge the fee and reward entirely. The most clean proposal seems to be to either just introduce signed integers as a class, or add some offset (eg. 2**63) to the combined variable.

Reply:
Bit the bullet and went with `int64` :)